Rewrite of vulnerability disclosure CS (#366)
OWASP/CheatSheetSeries: cheatsheets/Vulnerability_Disclosure_Cheat_Sheet.md

- Respond to reports in a reasonable timeline.
- Communicate openly with researchers.
- Not threaten legal action against researchers.
- Request CVEs where appropriate.
- Publish clear security advisories and changelogs.
- Offer rewards and credit.

## Contents

**FIXME**

## Methods of Disclosure

There are a number of different models that can be be followed when disclosing vulnerabilities, which are listed in the sections below.

### Private Disclosure

In the private disclosure model, the vulnerability is reported privately to the organisation. The organisation may choose to publish the details of the vulnerabilities, but this is done at the discretion of the organisation, not the researcher, meaning that many vulnerabilities may never be made public. The majority of bug bounty programs require that the researcher follows this model.

The main problem with this model is that if the vendor is unresponsive, or decides not to fix the vulnerability, then the details may never be made public. Historically this has lead to researchers getting fed up with companies ignoring and trying to hide vulnerabilities, leading them to the full disclosure approach.

### Full Disclosure

> With the full disclosure approach, the full details of the vulnerability are made public as soon as they are identified. This avoids the problem of organisations ignoring vulnerabilities, however it also means that the full details (sometimes including exploit code) are available to attackers, often before a patch is available.

Author: I've made some changes in 2b76608 - are you happy with these?

- A [security.txt](https://securitytxt.org) file on the website at `/security.txt` or `/.well-known/security.txt`.
- An existing issue tracking system.
- Generic email addresses such as `security@` or `abuse@`.
- The generic "Contact Us" page on the website.
- Social media platforms.
- Phoning the organisation.
- Reaching out to the community.

When reaching out to people who are not dedicated security contacts, request the details for a relevant member of staff, rather than disclosing the vulnerability details to whoever accepts the initial contact (especially over social media).

If it is not possible to contact the organisation directly, a national or sector-based  [CERT](https://en.wikipedia.org/wiki/Computer_emergency_response_team) may be able to assist.

### Initial Report

Once a security contact has been identified, an initial report should be made of the details of the vulnerability. Ideally this should be done over an encrypted channel (such as the use of PGP keys), although many organisations do not support this.

The initial report should include:

- Sufficient details of the vulnerability to allow it to be understood and reproduced.

> - HTML snippets, screenshots or any other supporting evidence.

Author: I've made some changes in 2b76608 - are you happy with these?

At a minimum, the security advisory must contain:

- A high level summary of the vulnerability, including the impact.
- A clear list of vulnerable versions.
- A clear list of patch versions.
- Any caveats on when the software is vulnerable (for example, if only certain configurations are affected).
- Any workarounds or mitigation that can be implemented as a temporary fix.
- A [CVE](https://cve.mitre.org/cve/request_id.html) for the vulnerability.

Where possible it is also good to include:

- The timeline of the vulnerability disclosure process.
- Credit for the researcher who identified the vulnerability.
- Technical details of the vulnerability.
- IDS/IPS signatures or other indicators of compromise.

Security advisories should be easy for developers and system administrators to find. Common ways to publish them include:

- A dedicated "security" or "security advisories" page on the website.
- A security mailing list or forum.
- Linked from the main changelogs and release notes.

> Some researchers may publish their own technical write ups of the vulnerability, which will usually include the full details required to exploit it (and sometimes even working exploit code). For more serious vulnerabilities, it may be sensible to ask the researcher to delay publishing the full details for a period of time (such as a week), in order to give system administrators more time to install the patches before exploit code is available. However, once the patch has been releases, attackers will be able to reverse engineer the vulnerability and developer their own exploit code, so there is limited value to delaying the full release.

Reviewer: ```suggestion
Some researchers may publish their own technical write ups of the vulnerability, which will usually include the full details required to exploit it (and sometimes even working exploit code). For more serious vulnerabilities, it may be sensible to ask the researcher to delay publishing the full details for a period of time (such as a week), in order to give system administrators more time to install the patches before any exploit code is available. However, once the patch has been released, attackers will be able to reverse engineer the vulnerability and develope their own exploit code, so there is limited value to delaying the full release.
```

### When to Give Up

Despite every effort that you make, some organisations are not interested in security, are impossible to contact, or may be actively hostile to researchers disclosing vulnerabilities. In some cases they may even threaten to take legal action against researchers. When this happens it is very disheartening for the researcher - it is important not to take this personally. When this happens, there are a number of options that can be taken.

- Publicly disclose the vulnerability, and deal with any negative reaction and potentially even a lawsuit. Whether or not they have a strong legal case is irrelevant - they have expensive lawyers and fighting any kind of legal action is expensive and time consuming. Before going down this route, ask yourself _is it really worth it?_

> - Anonymously disclose the vulnerability. However, if you've already made contact with the organisation and tried to report the vulnerability to them, it may be pretty obvious who you was responsible when it was anonymously disclosed. If you are going to take this approach, ensure that you have take sufficient operational security measures to protect yourself.

Reviewer: ```suggestion
- Anonymously disclose the vulnerability. However, if you've already made contact with the organisation and tried to report the vulnerability to them, it may be pretty obvious who's responsible behind the disclosure. If you are going to take this approach, ensure that you have taken sufficient operational security measures to protect yourself.
```
